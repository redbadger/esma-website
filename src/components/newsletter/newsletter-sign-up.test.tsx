import React from "react";
import { create } from "react-test-renderer";
import NewsletterSignUp from "./newsletter-sign-up";
import "@testing-library/jest-dom";
import addToMailchimp from "gatsby-plugin-mailchimp";
import { render, act, fireEvent, waitFor } from "@testing-library/react";

jest.mock("gatsby-plugin-mailchimp");

describe("NewsletterSignUp", () => {
  it("renders correctly", () => {
    const component = create(<NewsletterSignUp />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("hides submission errors on render", async () => {
    const { getByText } = render(<NewsletterSignUp />);

    expect(getByText("Please enter your name.")).not.toBeVisible();
    expect(getByText("Please enter your email.")).not.toBeVisible();
  });

  it("displays name error message when blank email is submitted", async () => {
    const { container, getByText } = render(<NewsletterSignUp />);

    act(() => {
      fireEvent.click(
        container.querySelector("input[type=submit][value=Submit]")
      );
    });
    expect(getByText("Please enter your name.")).toBeVisible();
  });

  it("displays email error message when blank email is submitted", async () => {
    const { container, getByText } = render(<NewsletterSignUp />);

    act(() => {
      fireEvent.click(
        container.querySelector("input[type=submit][value=Submit]")
      );
    });
    expect(getByText("Please enter your email.")).toBeVisible();
  });

  it("displays confirmation message if subscription was successful", async () => {
    const { container, getByText } = render(<NewsletterSignUp />);

    const promise = new Promise(resolve =>
      resolve({ result: "success", msg: "Thank you for subscribing" })
    );

    const name = "Stan Smith";
    const email = "stan.smith@example.org";

    addToMailchimp.mockImplementation(() => {
      return promise;
    });

    const nameInput = container.querySelector("input[type=text]");
    const emailInput = container.querySelector("input[type=email]");

    act(() => {
      fireEvent.change(nameInput, {
        target: { value: name },
      });
    });

    act(() => {
      fireEvent.change(emailInput, {
        target: { value: email },
      });
    });

    act(() => {
      fireEvent.click(
        container.querySelector("input[type=submit][value=Submit]")
      );
    });

    expect(addToMailchimp).toHaveBeenCalledWith(email, { FNAME: name });
    await waitFor(() => getByText("Thank you for subscribing!"));
  });

  it("warns the user if the email is already subscribed", async () => {
    const { container, getByText } = render(<NewsletterSignUp />);

    const promise = new Promise(resolve =>
      resolve({ result: "error", msg: "Email already subscribed." })
    );

    const name = "Stan Smith";
    const email = "stan.smith@example.org";

    addToMailchimp.mockImplementation(() => {
      return promise;
    });

    const nameInput = container.querySelector("input[type=text]");
    const emailInput = container.querySelector("input[type=email]");

    act(() => {
      fireEvent.change(nameInput, {
        target: { value: name },
      });
    });

    act(() => {
      fireEvent.change(emailInput, {
        target: { value: email },
      });
    });

    act(() => {
      fireEvent.click(
        container.querySelector("input[type=submit][value=Submit]")
      );
    });

    expect(addToMailchimp).toHaveBeenCalledWith(email, { FNAME: name });
    await waitFor(() => getByText("This email has already been subscribed."));
  });

  it("displays generic error message if something went wrong", async () => {
    const { container, getByText } = render(<NewsletterSignUp />);

    const promise = new Promise(resolve =>
      resolve({ result: "error", msg: "Some other error." })
    );

    const name = "Stan Smith";
    const email = "stan.smith@example.org";

    addToMailchimp.mockImplementation(() => {
      return promise;
    });

    const nameInput = container.querySelector("input[type=text]");
    const emailInput = container.querySelector("input[type=email]");

    act(() => {
      fireEvent.change(nameInput, {
        target: { value: name },
      });
    });

    act(() => {
      fireEvent.change(emailInput, {
        target: { value: email },
      });
    });

    act(() => {
      fireEvent.click(
        container.querySelector("input[type=submit][value=Submit]")
      );
    });

    expect(addToMailchimp).toHaveBeenCalledWith(email, { FNAME: name });
    await waitFor(() => getByText("Something went wrong, please try again later."));
  });
});
