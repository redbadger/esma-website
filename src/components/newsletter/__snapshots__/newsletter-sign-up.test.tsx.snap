// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsletterSignUp renders correctly 1`] = `
<section
  className="css-1x36vd0-newsletterSignUpStyles"
>
  <header>
    <h2>
      Join our mission
    </h2>
    <p>
      We’ve got bold ambitions and need impact focused leaders to join us on our journey. Sign up to our newsletter to be the first to hear about our events and stay up to date on our future plans.
    </p>
  </header>
  <article>
    <form
      onSubmit={[Function]}
    >
      <fieldset>
        <section
          className="css-usoz6a-newsletterInputFieldStyles"
        >
          <label
            htmlFor="fullNameInput"
          >
            Full name
          </label>
          <input
            className=""
            id="fullNameInput"
            name="fullName"
            onChange={[Function]}
            type="text"
          />
          <p
            className="input-error"
          >
            Please enter your name.
          </p>
        </section>
        <section
          className="css-usoz6a-newsletterInputFieldStyles"
        >
          <label
            htmlFor="emailInput"
          >
            Email address
          </label>
          <input
            className=""
            id="emailInput"
            name="email"
            onChange={[Function]}
            type="email"
          />
          <p
            className="input-error"
          >
            Please enter your email.
          </p>
        </section>
        <section
          className="submitSection"
        >
          <label
            htmlFor="submit"
          >
             
          </label>
          <input
            id="submit"
            type="submit"
            value="Submit"
          />
        </section>
        <section
          className="full-span"
        >
          <p
            className="response "
          />
          <p>
            By clicking Subscribe you accept the
             
            <a
              href="/privacy"
            >
              Privacy Policy
            </a>
            . This will allow us to send you newsletter and event updates. Our Privacy Policy covers how we protect data. You can unsubscribe from this at any point.
          </p>
        </section>
      </fieldset>
    </form>
  </article>
</section>
`;
